### PR TITLE
[WIP][ENH] TTM data pipeline: _dataset_class hook + TTMDataModule

### PR DESCRIPTION
**Pull request by StrikerEureka34**
#### Reference Issues/PRs
WIP towards #2184.  Related: #1959 #2051

#### What does this implement/fix? Explain your changes.

WIP draft for #2184, that is integrating IBM's TinyTimeMixer (TTM) into pytorch-forecasting's v2 pipeline. 
TTM is a pretrained foundation model for zero-shot and fine-tuned time series forecasting, and it expects a specific input format that our current v2 pipeline doesn't produce.

This draft/WIP PR covers the data layer only. No model weights, no `tsfm_public` dependency yet, as the main aim of it was to just getting the pipeline to speak TTM's language.

&ensp;

#### Issues I ran into

`TslibDataModule.setup()` had four hardcoded `_TslibDataset(...)` calls. 
So there was no clean way to swap in a different dataset class without duplicating the entire method.

&ensp;

#### Changes made

Added `_dataset_class = _TslibDataset` to TslibDataModule and replaced those four calls with `self._dataset_class(...)`. 
Which gives a zero behavioral change for existing code.

On top of that hook:

- _TTMDataset  now overrides `__getitem__` to produce TTM-format batches:
  past_values, past_observed_mask, future_values, prediction_channel_indices
- TTMDataModule now sets _dataset_class = _TTMDataset, derives channel indices
  from metadata, and provides a `collate_fn` for the mixed Tensor/non-Tensor batch

So the channel ordering follows TTM's spec:
`[targets | past-only covariates | known-future covariates]`

&ensp;

#### Does it work?

Yes, with a `MockTTM` standing in for the real model:

```
test_ttm_data_module_uses_ttm_dataset  PASSED
test_ttm_dataset_captures_stage        PASSED
test_channel_indices_derived           PASSED
test_batch_shapes                      PASSED
test_no_future_values_in_predict       PASSED
test_mock_ttm_forward                  PASSED
```

One pre-existing failure in test_multivariate_target ('list' object has no attribute 'shape') which was reproducible on `main`, so unrelated to this PR and the changes made by it, worth a separate issue if not already tracked?

&ensp;

#### What's left

- **Caching**: `_preprocess_data` runs per `__getitem__`  so it needs a per-series cache before this is usable at scale (I mentioned it as a TODO in the code and using a short term fix for current scope)
- **Model wrapper**: thin wrapper around `tsfm_public.models.tinytimemixer`, as an optional dependency.
- **Training glue**: loss, forward, predict steps wired through `TslibBaseModel`

&ensp;

The _dataset_class hook is intentionally generic, with TTMDataModule as the first user, but the same pattern should work for Chronos, Moirai, and others with their own `__getitem__` overrides.
More discussion and testing is needed for that I believe, but initial findings seem promising.

Looking for early feedback on this hook design and TTM channel format assumptions before going further, thanks!
@phoeenniixx @PranavBhatP @agobbifbk

### CODE REVIEW

**Comment by agobbifbk** [comment]
Are you sure that we need to dedicate a custom data layer for a single model? Now you are aware on what the model expects as input. I would like to see something like I posted [here](https://github.com/sktime/pytorch-forecasting/issues/2051#issuecomment-4053508328) where you try to figure out if the pieces (in terms of tensors, past, future, categorical, numerical) we have in the current data module can be rearranged in order to build the batch that TTM expects. If we can move the logic to the forward loop we can avoid custom data layer. If you see an high incompatibility I would like to see why you think it is not possible. THX

**Comment by StrikerEureka34** [comment]
> Are you sure that we need to dedicate a custom data layer for a single model? Now you are aware on what the model expects as input. I would like to see something like I posted [here](https://github.com/sktime/pytorch-forecasting/issues/2051#issuecomment-4053508328) where you try to figure out if the pieces (in terms of tensors, past, future, categorical, numerical) we have in the current data module can be rearranged in order to build the batch that TTM expects. If we can move the logic to the forward loop we can avoid custom data layer. If you see an high incompatibility I would like to see why you think it is not possible. THX

Yeah this makes a lot of sense. Two wins if we do it this way:
1. No need for a dedicated data layer for TTM as _TslibDataset already carries everything.
2. cats can go straight through nn.Embedding inside forward, which honestly wouldn't have worked in a worker-side dataset anyway (embeddings live on the main process). 

That second point I'd missed completely, so thanks for pushing back, I hope this is what you meant?

Untitled Diagram (1)

I've added a gist here with the test suite to check if the rearrangement works and if cats can go in as embeddings, while keeping the pointers and diagram you mentioned in #2051 in mind: [here](https://gist.github.com/StrikerEureka34/160d985257d1147b4c95013dd4158080). 

What the results mean:-
- `Part 1 max|forward - dataset|=0.00e+00`: 
Forward-loop is bit-identical to `_TTMDataset`. So nothing will be lost if we move it out.
- `Part 2 (20, 11) vs (20, 3) (diff = 8 cat-emb channels)`: 
The cats are carried by forward and silently dropped by `_TTMDataset`. 
min_emb_grad_abs_sum=80 confirms embeddings are learnable only when they sit on the model.
I had missed this completely, and its all the more reason why we don't need the separate data module here.
- `R1/R2/R3 all return True for both cont and cat..`

Some of the implications that I can see for this pr after the analysis are:-
- `_TTMDataset` + custom collate go.
- TTMDataModule goes too. TslibDataModule is used as-is; its metadata["feature_names"] already has continuous / categorical / known.
- `TTMForecaster.__init__` derives past_only_* / known_future_* splits from metadata once. forward uses them for rearrangement + nn.Embedding.
- `_dataset_class` hook stays available for Chronos/Moirai if they ever need dataset-side work just nobody overrides it for TTM, should we keep this for flexibility or remove it entirely? 

So what this would finally look like is:-

Untitled Diagram drawio (10)



Please let me know if this provides a clearer picture, or there are more options that I might've missed? 
Thanks! @agobbifbk @phoeenniixx

**Comment by agobbifbk** [comment]
Much better now, so you idea is that we can build the batch to be compliant with the TTM model, but there are some work to do:
1- ensure the order of the targets
2- have a clear idea if, as for chronos2, the future covariates needs to be also past covariates, in case this is mandatory and the user did not provide the correct metadata raise a warning and fix the past, or raise an error
3- if 2 implies that if you want to use the future covariates they should be also in the past, be sure that the order is correct

The good news is that the reordering part can be done BY DEFAULT since this has not implication in the other models (or at least it should). The subsetting (know should be a subset of the past) can be checked after model initialization using the metadata, isn't it?

If you have all the pieces you can start implementing TTM or raise an issue for doing it!
THX so much!
